### ISSUE TITLE
Clone application

### ISSUE THREAD

**Issue by CAFxX** [cloudfoundry/cli#1212]
## Command  

```
cf clone APP OLD_APP [--no-start] [--no-routes] [--no-services] [-o/--org OLD_ORG] [-s/--space OLD_SPACE]
```

## What occurred

currently not supported

## What you expected to occur

A new `APP` is created with the same bits and configuration of `OLD_APP` (optionally from `OLD_SPACE`/`OLD_ORG`, assuming the user has access to them). If `--no-start` is specified, the application is only cloned but not staged/started. If `--no-routes` is specified, the routes of `OLD_APP` are not mapped to `APP`. If `--no-services` is specified, the services of `OLD_APP` are not bound to `APP`.

If cloning across spaces/orgs and a route/service is not available in the destination space/org the command should fail with a tip to run it with --no-services/--no-routes.

## CLI Version  

cf version 6.29.2+c66d0f3.2017-08-25

## CC API Endpoint Version  

2.74.0

## Any other relevant information  

Command naming can be bikeshedded as much as required (cf clone, cf copy, cf clone-app, ...)

**Comment by dkoper** [comment]
Hi @CAFxX 

Would the user's expectation be that this command uses the same app bits, or the same droplet (which ensures same buildpack version, stack, etc.) is used?

**Comment by CAFxX** [comment]
For the primary use case I had in mind it would be same bits, to improve the restage process with new buildpack versions. 

I can also see some use cases for cloning the droplet, maybe I can open a separate ticket for that?

**Comment by dkoper** [comment]
For the droplet, rather than the CLI having to download a droplet from one space and uploading it to another, it may make more sense for CC to support an operation for that.

**Comment by dkoper** [comment]
Btw, you're asking for a solution (a clone command). Can you explain more about the problem it is designed to solve?

**Comment by CAFxX** [comment]
> Can you explain more about the problem it is designed to solve?

Sure! The goal is to not require users to do a cf push in the following scenarios (assume the application is in production):

- I want to change an environment variable without stopping the application (clone + b/g)
- I want to bind/unbind a service without stopping the application (clone + b/g)
- I want to change (upgrade) the buildpack version without stopping the application (clone + b/g)
- I want to troubleshoot/reproduce a bug without impacting live traffic: I clone the application with a different route and I can use it "safely" for e.g. attaching a debugger

This will become more and more important with the fine-grained auth effort that is currently underway. Many of our users have big teams and they don't want anyone to be able to push bits (but would be fine with making a copy of an existing app).

There may be other use cases, I will update the ticket in that case.

> For the droplet, rather than the CLI having to download a droplet from one space and uploading it to another, it may make more sense for CC to support an operation for that.

Yup, that was one of the reasons this ticket is about cloning the bits and not the droplet. I know copy the bits from app to app is supported by CC, but AFAIK copying a droplet from app to app is not.

**Comment by CAFxX** [comment]
@dkoper I updated the first post with the use cases

**Comment by mohanbabubb** [comment]
Anyone seeing this!

**Comment by abbyachau** [comment]
Hi @CAFxX thanks for this and for folks who have commented. I've spoken to you before about a new rolling deployments feature that the CC API are currently implementing. Please see this [feature narrative](https://docs.google.com/document/d/116I_mOWjZcPeIbUvvsh-jAcwpoE_mGPD_SkCel5xXuU/edit).  It sounds like rolling deployments would resolve some of the use cases you've mentioned above. Is that statement accurate? What would rolling deployments not solve for your use cases?

cc @tcdowney @charleshansen

**Comment by CAFxX** [comment]
@abbyachau yes, zero-downtime push would solve some of the same problems; AFAICT it doesn't address:

- promoting a pushed app from space A to space B, when there is a need to guarantee that the bits or bits+droplet are the same (in the latter case, assuming stacks are the same)
- allowing users to keep original app and cloned app running side-by-side, especially useful when troubleshooting or for canary deploys

Note that both of the things above can be done today with the current APIs, so this is merely a CLI issue.

**Comment by abbyachau** [comment]
cc @charleshansen @tcdowney

**Comment by jenspinney** [comment]
These seem like important use cases. I know you're asking for a new command for this, but as a workaround, the steps @tjvman mentioned here might help: https://github.com/cloudfoundry/cli/issues/1498#issuecomment-444211657. These steps don't address the use case of wanting to experiment with a new/different buildpack, but it may help with some of the scenarios you mentioned. Not exactly what you're looking for, but maybe it'll help some who stumble upon this GitHub issue.

**Comment by mamachanko** [comment]
pardon that I haven't read the whole thread in detail yet, but could [`cf copy-source`](https://cli.cloudfoundry.org/en-US/cf/copy-source.html) help?

**Comment by gururajsh** [comment]
Hi @CAFxX ,

At this time, we will not be able to pursue this. If this issue is still relevant, please open a new GH Issue.
